### PR TITLE
ctranslate2: enable cuDNN only if it is available

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

Guard use of `cudnn` with its existence in the package set -- it's not always available (e.g., Jetson or certain CUDA releases).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20931,7 +20931,8 @@ with pkgs;
   ctranslate2 = callPackage ../development/libraries/ctranslate2 rec {
     stdenv = if withCUDA then gcc11Stdenv else pkgs.stdenv;
     withCUDA = pkgs.config.cudaSupport;
-    withCuDNN = pkgs.config.cudaSupport;
+    withCuDNN = withCUDA && (cudaPackages ? cudnn);
```
Comment (SomeoneSerge) [inline pkgs/top-level/all-packages.nix:20934]: The definition is going to need a `meta.broken = !(withCuda -> cudaPackages ? cudnn)`

File: pkgs/development/libraries/ctranslate2/default.nix [hunk 2]
```diff
@@ -87,5 +87,6 @@ stdenv.mkDerivation rec {
     changelog = "https://github.com/OpenNMT/CTranslate2/blob/${src.rev}/CHANGELOG.md";
     license = licenses.mit;
     maintainers = with maintainers; [ hexa misuzu ];
+    broken = !(withCuDNN -> withCUDA);
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/ctranslate2/default.nix:90]: `broken = false` when we manually pass `withCUDA = withCuDNN = true` and `cudaPackages_10`?

Comment (ConnorBaker) [reply pkgs/development/libraries/ctranslate2/default.nix:90]: `ctranslate` uses CUDA redistributables and doesn't appear to support the runfile installer, so it won't eval pre-11.4. I'll add a guard for that

Comment (ConnorBaker) [reply pkgs/development/libraries/ctranslate2/default.nix]: Resolved in latest; shouldn't break eval now.

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 272084 --extra-nixpkgs-config '{
      allowUnfree = true;
      allowBroken = false;
      cudaSupport = true;
      cudaCapabilities = [ "8.9" ];
    }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Review by SomeoneSerge — APPROVED** [review]

**Pull request merged**